marc.py

```diff
@@ -0,0 +1,612 @@
+    def create_record(cls, work, annotator, force_create=False, integration=None):
+        """Build a complete MARC record for a given work."""
+        if callable(annotator):
+            annotator = annotator()
+
+        if isinstance(work, BaseMaterializedWork):
+            pool = work.license_pool
+        else:
+            pool = work.active_license_pool()
+        if not pool:
+            return None
+
+        edition = pool.presentation_edition
+        identifier = pool.identifier
+
+        _db = Session.object_session(work)
+
+        record = None
+        existing_record = getattr(work, annotator.marc_cache_field)
+        if existing_record and not force_create:
+            record = Record(data=existing_record, force_utf8=True)
+
+        if not record:
+            record = Record(leader=annotator.leader(work), force_utf8=True)
+            annotator.add_control_fields(record, identifier, pool, edition)
+            annotator.add_isbn(record, identifier)
+
+            # TODO: The 240 and 130 fields are for translated works, so they can be grouped even
+            # though they have different titles. We do not group editions of the same work in
+            # different languages, so we can't use those yet.
+
+            annotator.add_title(record, edition)
+            annotator.add_contributors(record, edition)
+            annotator.add_publisher(record, edition)
+            annotator.add_physical_description(record, edition)
+            annotator.add_audience(record, work)
+            annotator.add_series(record, edition)
+            annotator.add_system_details(record)
+            annotator.add_ebooks_subject(record)
+
+
+            data = record.as_marc()
+            if isinstance(work, BaseMaterializedWork):
+                setattr(pool.work, annotator.marc_cache_field, data)
+            else:
+                setattr(work, annotator.marc_cache_field, data)
+
+        # Add additional fields that should not be cached.
+        annotator.annotate_work_record(work, pool, edition, identifier, record, integration)
+
+        return record
+
+    def records(self, lane, annotator, force_refresh=False, mirror=None):
+        output = StringIO()
+
+        works_q = lane.works(self._db)
+        for work in works_q:
+            record = self.create_record(
+                work, annotator, force_refresh, self.integration)
+            if record:
+                output.write(record.as_marc())
+
+        content = output.getvalue()
+        output.close()
+
+        # Mirror the content.
+        storage_protocol = self.integration.setting(self.STORAGE_PROTOCOL).value
+        if storage_protocol == ExternalIntegration.S3:
+            mirror_integration = ExternalIntegration.lookup(
+                self._db, storage_protocol, ExternalIntegration.STORAGE_GOAL)
+            if not mirror_integration:
+                raise Exception("No mirror integration for configured protocol %s" % storage_protocol)
```

Comment:
This is what prompted me to make the comment in https://github.com/NYPL-Simplified/server_core/pull/993/files#r238435014

This is pretty close to what happens in `MirrorIntegration.sitewide_integration`, and I don't think this way of writing it is more future-proof than checking a yes-or-no setting that says whether we should call `sitewide_integration`.

Reply:
Yes, I think a configuration setting makes sense. We would either need to add a column to ConfigurationSetting or deal with the edge case where an ExternalIntegration is deleted but its ID was referred to by ConfigurationSettings.

Reply:
I should explicitly say that I think switching to yes-or-no is a better solution for now.